apiserver/common/secrets/secrets.go

```diff
@@ -672,87 +672,136 @@ func RemoveSecretsForAgent(
 	return result, nil
 }
 
-// RemoveUserSecrets removes the specified user supplied secrets.
-// The secrets are removed from the state and backend, and the caller must have model admin access.
-func RemoveUserSecrets(
-	removeState SecretsRemoveState, adminConfigGetter BackendAdminConfigGetter,
-	authTag names.Tag, args params.DeleteSecretArgs,
-	modelUUID string,
-	canDelete func(*coresecrets.URI) error,
-) (params.ErrorResults, error) {
-	removeFromBackend := func(p provider.SecretBackendProvider, cfg provider.ModelBackendConfig, revs provider.SecretRevisions) error {
-		backend, err := p.NewBackend(&cfg)
+// removeRevisionFromBackend removes the specified revision from the specified secret backend.
+func removeRevisionFromBackend(backendCfg *provider.ModelBackendConfig, revisionId string) error {
+	p, err := GetProvider(backendCfg.BackendType)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// delete secret in backend
+	backend, err := p.NewBackend(backendCfg)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = backend.DeleteContent(context.TODO(), revisionId); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// removeSecretFromExternal removes the specified secret from its respective secret backends.
+// Can be called for specific revisions, or omit revisions to remove all revisions of the secret.
+func removeSecretFromExternal(removeState SecretsRemoveState, adminConfigGetter BackendAdminConfigGetter, authTag names.Tag, uri *coresecrets.URI, revisions ...int) error {
+	var revs []*coresecrets.SecretRevisionMetadata
+	var err error
+	if len(revisions) == 0 {
+		// Remove all revisions.
+		revs, err = removeState.ListSecretRevisions(uri)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		for _, revId := range revs.RevisionIDs() {
-			if err = backend.DeleteContent(context.TODO(), revId); err != nil {
+	} else {
+		// Remove specified revisions.
+		revs = make([]*coresecrets.SecretRevisionMetadata, len(revisions))
+		for i, rev := range revisions {
+			revMeta, err := removeState.GetSecretRevision(uri, rev)
+			if err != nil {
 				return errors.Trace(err)
 			}
+			revs[i] = revMeta
```

Comment:
We don't want to record the entire revision metadata struct. We are just interested in the ValueRef, which might be nil if the secret is stored in mongo. We'll also want to check drain behaviour to ensure it can handle the race where external content is deleted during a drain and revision metadata still exists.

Reply:
Can you take another look after the recent changes?  The code below uses:
1.* `rev.ValueRef` (like you mentioned)
2. `rev.Revision` (used deletion fails with `NotFound`, so we can check if the revision was drained elsewhere
3. `rev.BackendName` (just for some error annotation)

(3) is non functional and could be removed, but we still need `rev.Revision` at least.  I could distill these down to a simpler type instead of `[]*secrets.SecretRevisionMetadata` if you'd prefer